src/core/geocms/qgsgeonoderequest.h

```diff
@@ -0,0 +1,100 @@
+/***************************************************************************
+    qgsgeonoderequest.h
+    ---------------------
+    begin                : Jul 2017
+    copyright            : (C) 2017 by Muhammad Yarjuna Rohmat, Ismail Sunni
+    email                : rohmat at kartoza dot com, ismail at kartoza dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+#ifndef QGSGEONODEREQUEST_H
+#define QGSGEONODEREQUEST_H
+
+#include "qgis.h"
+#include "qgis_core.h"
+#include <qnetworkreply.h>
+
+
+#include <QObject>
+#include <QUuid>
+
+struct LayerStruct
+{
+  QUuid uuid;
+  QString name;
+  QString typeName;
+  QString title;
+  QString wmsURL;
+  QString wfsURL;
+  QString xyzURL;
+};
+
+class CORE_EXPORT QgsGeoNodeRequest : public QObject
+{
+    Q_OBJECT
+  public:
+    explicit QgsGeoNodeRequest( bool forceRefresh, QObject *parent = nullptr );
+    QgsGeoNodeRequest( const QString &baseUrl, /*const QgsWmsAuthorization &auth,*/ bool forceRefresh, QObject *parent = nullptr );
+    virtual ~QgsGeoNodeRequest();
+
+    bool getLayers();
+
+    QList<LayerStruct> parseLayers( QByteArray layerResponse );
+
+    // Obtain list of unique URL in the geonode
+    QStringList serviceUrls( QString serviceType );
+
+    // Obtain map of layer name and url for a service type
+    QgsStringMap serviceUrlData( QString serviceType );
+
+    QString lastError() const { return mError; }
+
+    QByteArray response() const { return mHttpGeoNodeResponse; }
+
+    QNetworkReply *reply() const { return mGeoNodeReply; }
+
+    //! Abort network request immediately
+    void abort();
+
```

Comment:
There are various functions in the public part of the API that probably do not need to be there. Which methods are called from outside?

I am also wondering about all the networking code in the class - is it maybe possible to replace all that code by QgsNetworkContentFetcher class? The big disadvantage of the current code is that it is made to run synchronously, which will block the user interface while the request is being processed, which is bad for user experience.

Ideally I would hope we do not need this class in core library at all. From what I understand, it does three things: 1. fetch data from server, 2. parse data into the LayerStruct structure, 3. provides convenience methods serviceUrls(), serviceUrlData().

In my opinion, the best approach would be to remove this class and move its key functionality to QgsGeoNodeConnection class as static methods:
1. static method to get URL to get capabilities, which client code can later use with QgsNetworkContentFetcher to download data
2. static method to parse layer details from QByteArray with server's capabilities
3. static methods serviceUrls(), serviceUrlData() to extract data from a list of layer details.